sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/util/Pair.java

```diff
@@ -28,7 +29,7 @@
  * @param <V> the type of the second element of the pair
  */
 @Audience(Audience.Type.CLIENT)
-public final class Pair<K, V> {
+public final class Pair<K, V> implements Serializable {
```

Comment:
I think this should be
```
public final class Pair<K extends Serializable, V extends Serializable> implements Serializable
```
otherwise the `implements Serializable` in invalid.

Reply:
I see your point. Generally speaking what is a meaning of having `serializable` container when items within it are not. 

Small trouble is that enforcing `Pair<K extends Serializable, V extends Serializable> ` brings up 70 compilation errors. Those may bring another `implements Serialization` enforcements, for example keys in `ReduceByKey` and `Join` operators, and some code rewrites to avoid it.

Reply:
Hm, that is definitely not intended to force keys in R(S)BK to be Serializable. That would be a major breaking change and mostly use-less. Maybe we could make other change - the Serializable is needed only in certain situations. Maybe we could create private static subclass of `Pair`, which would extend it and be serializable (with restrictions to the type parameters) and add additional method
```
  public static <K extends Serializable, V extends Serializable> Pair<K, V> ofSerializable(K k, V v)
```
and use it wherever we need to force the object to be Serializable?

Reply:
We would probably have to use custom serialization, because the parent Pair class of the derived class is not serializable.

Reply:
Another option might be to turn the Pair into interface.

Reply:
I see your point but I don't like the idea of forcing user to have serializable data. Beam's KV is also serializable and does not enforce K, V to be serializable.

Reply:
I'd say only use case for this are tests and "Create" transformation, which will enforce user to make data serializable anyway

Reply:
Yes, the problem is that Pair is used internally for keyed operations. This definitely means, that we don't want the `K`, `V` to be Serializable in all cases. Looks like I will have to learn how to live with this. :-)

Reply:
I includes 7de0ebc as a thought experiment. It allows to have both serializable `Pair` with and without type parameters enforced to be serializable. Please review and comment.